src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -83,85 +70,37 @@ public ArrayHandler([CanBeNull] NpgsqlTypeHandler elementHandler, int lowerBound
-
-            // Getting an array
-            if (t.IsArray)
-            {
-                var elementType = t.GetElementType();
-                // We need to treat this as an actual array type, these need special treatment because of
-                // typing/generics reasons (there is no way to express "array of X" with generics
-                if (elementType == GetElementFieldType())
-                    return (TArray)await ReadAsObject(buf, len, async, fieldDescription);
-                if (elementType == GetElementPsvType())
-                    return (TArray)await ReadPsvAsObject(buf, len, async, fieldDescription);
-            }
-
-            if (t == typeof(List<TElement>))
-                return (TArray)await ReadAsList<TElement>(buf, async);
-
-            if (t.IsGenericType
-                && t.GetGenericTypeDefinition() == typeof(List<>)
-                && t.GenericTypeArguments[0] == GetElementPsvType())
-            {
-                return (TArray)await ReadPsvAsList(buf, async);
-            }
-
-            throw new InvalidCastException($"Can't cast database type {PgDisplayName} to {typeof(TArray).Name}");
-        }
+        public override ValueTask<Array> Read(NpgsqlReadBuffer buf, int len, bool async, FieldDescription fieldDescription = null)
+            => ReadArray<TElement>(buf, async);
 
-        protected async ValueTask<object> ReadAsList<TElement2>(NpgsqlReadBuffer buf, bool async)
+        protected internal override ValueTask<TAny> Read<TAny>(NpgsqlReadBuffer buf, int len, bool async, FieldDescription fieldDescription = null)
         {
-            await buf.Ensure(12, async);
-            var dimensions = buf.ReadInt32();
+            if (typeof(TAny) == typeof(List<TElement>))
+                return ReadListImpl<TAny>(buf, async);
 
-            if (dimensions == 0)
-                return new List<TElement2>();
-            if (dimensions > 1)
-                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(TElement2).Name}>");
+            if (IsArrayOf<TAny, TElement>.Value)
+                return ReadArrayImpl<TAny>(buf, async);
 
-            buf.ReadInt32();  // Has nulls. Not populated by PG?
-            buf.ReadUInt32(); // Element OID. Ignored.
-
-            await buf.Ensure(8, async);
-            var length = buf.ReadInt32();
-            buf.ReadInt32(); // We don't care about the lower bounds
-
-            var list = new List<TElement2>(length);
-            for (var i = 0; i < length; i++)
-                list.Add(await ElementHandler.ReadWithLength<TElement2>(buf, async));
-            return list;
+            return base.Read<TAny>(buf, len, async, fieldDescription);
         }
 
-        protected internal virtual ValueTask<object> ReadPsvAsList(NpgsqlReadBuffer buf, bool async)
-            => ReadAsList<TElement>(buf, async);
+        async ValueTask<TArray> ReadArrayImpl<TArray>(NpgsqlReadBuffer buf, bool async)
```

Comment:
I agree with @roji - they're one-liners after all.

Reply:
Because of the base method call in the last return statement in `Read<TAny>`.

Reply:
Is this an attempt to keep `Read<TAny>` non-async by moving the actual async logic into local functions, like I did in several places with Long methods ([this is an example](https://github.com/npgsql/npgsql/blob/183be594fafb388c984411f8e6f61ac1a3870596/src/Npgsql/NpgsqlDataReader.cs#L1750)? If so, the general idea is nice but I don't see the usefulness here, because in all normal executions of `Read<TAny>` one of the async one-liner local functions will be called anyway. The only flow in which we don't call into one of the local methods is if the `TAny` is neither an array, nor a list, in which case we call `base.Read<TAny>()` just so that the proper exception gets thrown (by the way that's a great idea).

Optimizing to avoid making `Read<TAny>` just for the error condition doesn't make much sense to me... The only places where I do this kind of non-async/async optimization is where the non-async path is a real, critical fast path, where there's a good chance data is already in memory, and we don't want to pay the overhead of an async state machine etc. 

To summarize, I don't see a reason for `Read<TAny>` (and the other equivalent examples) not to look like the following code. But @YohDeadfall maybe I'm missing something or not understanding...

```c#
protected internal override async ValueTask<TAny> Read<TAny>(NpgsqlReadBuffer buf, int len, bool async, FieldDescription fieldDescription = null)
{
    if (IsArrayOf<TAny, TElement>.Value)
        return (TAny)(object)await ReadArray<TElement>(buf, async);

    if (typeof(TAny) == typeof(List<TElement>))
        return (TAny)(object)await ReadList<TElement>(buf, async);

    return await base.Read<TAny>(buf, len, async, fieldDescription);
}
```

Reply:
Okay, will do, but before that please answer my question below.